### ISSUE TITLE
Organise devices in Folders

### ISSUE THREAD

**Issue by PierreScerri** [esphome/feature-requests#867]
**Describe the problem you have/What new integration you would like**

It would be nice and more efficient to organise my ESPHome devices into folders.

**Please describe your use case for this integration and alternatives you've tried:**

As the number of ESPHome devices in my home increases, it is becoming more tedious to find a specific device.

One could also possibly use the 'Update' option on a folder instead of a 'One or All' update. 

'Update all' is not really useful if you have updated an !include file for a bunch sonoff duals and you only need to update those and not all the other Basics, Minis and Wemoses.

Having a grid display instead of a long list would also help.

Thank you

**Additional context**

--

**Comment by thedexboy** [comment]
I've installed the dev version yesterday and it has a grid layout. Way better than before.

**Comment by PierreScerri** [comment]
My ESPHome is an addon (ver 1.14.5) to Hassio. I don't think that it gives me the option to install the dev version.

**Comment by glmnet** [comment]
Yes, beta and dev, both have the new feature and both are available as add-ons, you can install them, check docs page https://esphome.io/guides/faq.html#how-do-i-update-to-the-latest-beta-release

**Comment by PierreScerri** [comment]
That is already a lot better.  Thank you.

**Comment by glmnet** [comment]
If somebody is going to work on this, I believe the folder should not be virtual but follow the filesystem folder structure, this will make more sense as you can use packages / fonts / secrets, etc. and include files in other folders.

**Comment by jbuszkie** [comment]
I hope someone can work on it... 🤞

**Comment by lcavalli** [comment]
> If somebody is going to work on this, I believe the folder should not be virtual but follow the filesystem folder structure, this will make more sense as you can use packages / fonts / secrets, etc. and include files in other folders.

I agree. Having support for sub-folders at filesystem level it will be possible to archive additional files alongside the yaml configuration, like PCB gerber files, pdf with protocol specification, .gitignore files, additional .h and .cpp source code without messing up with other nodes.

**Comment by maslyankov** [comment]
Any news on this topic?

**Comment by sjpilot** [comment]
A partial workaround is to use the !include yaml per https://esphome.io/guides/configuration-types.html#yaml-insertion-operator. Unfortunately you will still need a file per device but it can be a placeholder to include a file in a git repository folder.
`<<: !include testproject/testdevice.yaml
`

**Comment by yonz2** [comment]
I'd like to add a concrete downstream use case for this, plus a specific technical detail on why the current workaround is more painful than it needs to be.

**Use case**: running the Dashboard centrally for a multi-tenant teaching lab — one Dashboard instance, but per-classroom/"Area" isolation for both credentials and device listing. Real filesystem subdirectories per Area (as @glmnet described above) would be the natural fit, since it'd let each Area also keep its own secrets.yaml, package files, etc. alongside its devices, the same way a single-tenant user would want packages/fonts/secrets alongside theirs.

What we found testing this today: the Dashboard's device scanner is non-recursive — confirmed against the live source (esphome/dashboard/dashboard.py's _list_dashboard_entries, which goes through esphome.util.list_yaml_files() → config_dir.glob("*"), a single-level glob with no recursion). 
A device YAML placed in any subdirectory of /config is silently never discovered at all — it doesn't error, it just doesn't appear in the list. That's a step below "no folders in the UI" — it's "folders don't work as a workaround either," which pushed us toward a flatter, prefix-based naming scheme instead (e.g. area01-devicename.yaml) purely to fake namespacing.

Given @glmnet's comment already settled on filesystem-backed folders as the right shape, would a scoped first step of "make list_yaml_files() recursive (one level, or fully recursive with glob("**/*.yaml")), and display the relative path as the device's grouping in the UI" be a reasonable, smaller PR to attempt independently of the fuller folder-management UI this issue also asks for?

**Comment by emandtf** [comment]
It could be useful even in a standalone PY/Docker installation (I mean without HA/HASSIO).

I have more than few projects that uses "_includes: ..._" of external C/C++ files which are placed in a directory at the same level of project YAML.
I would like to do a "GIT Revisioning" of ***each*** project but I cannot do "_git init_" from ESPHome's project root (to initialize ONE project repository) to include one YAML and relative folder for each project.......

At this time I have to copy YAML file inside the relative subfolder and then do the Commit/Push. To restore the Commit I have to not forget to move YAML to parent directory after each "git clone/pull".

It's a bit tricky......
